lib/postgres.go

```diff
@@ -0,0 +1,741 @@
+	if err != nil {
+		return nil, err
+	}
+
+	blockMap := make(map[BlockHash]*BlockNode)
+	for _, block := range blocks {
+		blockMap[*block.Hash] = &BlockNode{
+			Hash:             block.Hash,
+			Height:           uint32(block.Height),
+			DifficultyTarget: block.DifficultyTarget,
+			CumWork:          HashToBigint(block.CumWork),
+			Header: &MsgBitCloutHeader{
+				Version:               block.Version,
+				PrevBlockHash:         block.ParentHash,
+				TransactionMerkleRoot: block.TxMerkleRoot,
+				TstampSecs:            block.Timestamp,
+				Height:                block.Height,
+				Nonce:                 block.Nonce,
+				ExtraNonce:            block.ExtraNonce,
+			},
+			Status: block.Status,
+		}
+	}
+
+	// Setup parent pointers
+	for _, blockNode := range blockMap {
+		// Genesis block has nil parent
+		parentHash := blockNode.Header.PrevBlockHash
+		if parentHash != nil {
+			blockNode.Parent = blockMap[*parentHash]
+		}
+	}
+
+	return blockMap, nil
+}
+
+func (postgres *Postgres) GetChain(name string) *Chain {
+	chain := &Chain{
+		Name: name,
+	}
+
+	err := postgres.db.Model(chain).First()
+	if err != nil {
+		return nil
+	}
+
+	return chain
+}
+
+func (postgres *Postgres) UpsertChain(name string, tipHash *BlockHash) error {
+	return postgres.db.RunInTransaction(postgres.db.Context(), func(tx *pg.Tx) error {
+		return postgres.UpsertChainTx(tx, name, tipHash)
+	})
+}
+
+func (postgres *Postgres) UpsertChainTx(tx *pg.Tx, name string, tipHash *BlockHash) error {
+	bestChain := &Chain{
+		TipHash: tipHash,
+		Name:    name,
+	}
+
+	_, err := tx.Model(bestChain).WherePK().OnConflict("(name) DO UPDATE").Insert()
+	return err
+}
+
+func (postgres *Postgres) InsertTransactionsTx(tx *pg.Tx, bitCloutTxns []*MsgBitCloutTxn, blockHash *BlockHash) error {
+	var transactions []*Transaction
+	var transactionOutputs []*TransactionOutput
+	var transactionInputs []*TransactionOutput
+	var metadataBlockRewards []*MetadataBlockReward
+	var metadataBitcoinExchanges []*MetadataBitcoinExchange
+	var metadataPrivateMessages []*MetadataPrivateMessage
+	var metadataSubmitPosts []*MetadataSubmitPost
+	var metadataUpdateProfiles []*MetadataUpdateProfile
+	var metadataExchangeRates []*MetadataUpdateExchangeRate
+	var metadataFollows []*MetadataFollow
+	var metadataLikes []*MetadataLike
+	var metadataCreatorCoins []*MetadataCreatorCoin
+	var metadataCreatorCoinTransfers []*MetadataCreatorCoinTransfer
+	var metadataSwapIdentities []*MetadataSwapIdentity
+
+	for _, txn := range bitCloutTxns {
+		txnHash := txn.Hash()
+		transaction := &Transaction{
+			Hash:      txnHash,
+			BlockHash: blockHash,
+			Type:      txn.TxnMeta.GetTxnType(),
+			PublicKey: txn.PublicKey,
+			ExtraData: txn.ExtraData,
+		}
+
+		if txn.Signature != nil {
+			transaction.R = BigintToHash(txn.Signature.R)
+			transaction.S = BigintToHash(txn.Signature.S)
+		}
+
+		transactions = append(transactions, transaction)
+
+		for i, input := range txn.TxInputs {
```

Comment:
ii

Reply:
To clarify, Mr. Hands does not like using single letter variables in code because they are hard to find and replace.  We should replace all instances of "for i".